options/configurable.cc

```diff
@@ -0,0 +1,971 @@
+    if (opts_str.find(';') == std::string::npos &&
+        opts_str.find('=') == std::string::npos) {
+      return ParseStringOptions(db_opts, opts_str, input_strings_escaped,
+                                ignore_unused_options);
+    } else {
+#ifndef ROCKSDB_LITE
+      std::unordered_map<std::string, std::string> opt_map;
+      std::unordered_set<std::string> unused_opts;
+
+      s = StringToMap(opts_str, &opt_map);
+      if (s.ok()) {
+        s = DoConfigureFromMap(db_opts, opt_map, input_strings_escaped,
+                               ignore_unused_options, &unused_opts);
+      }
+#else
+      s = ParseStringOptions(db_opts, opts_str, input_strings_escaped,
+                             ignore_unused_options);
+#endif  // ROCKSDB_LITE
+    }
+  }
+  return s;
+}
+
+/**
+ * Sets the value of the named property to the input value, returning OK on
+ * succcess.
+ */
+Status Configurable::ConfigureOption(const DBOptions& db_opts,
+                                     const std::string& name,
+                                     const std::string& value,
+                                     bool input_strings_escaped) {
+  const std::string& option_name = GetOptionName(name);
+  const std::string& option_value =
+      input_strings_escaped ? UnescapeOptionString(value) : value;
+  bool found_it = false;
+  Status status;
+#ifndef ROCKSDB_LITE
+  status = SetOption(db_opts, option_name, option_value, input_strings_escaped,
+                     &found_it);
+#endif
+  if (!found_it) {
+    status = SetUnknown(db_opts, option_name, option_value);
+  }
+  return status;
+}
+
+/**
+ * Looks for the named option amongst the options for this type and sets
+ * the value for it to be the input value.
+ * If the name was found, found_option will be set to true and the resulting
+ * status should be returned.
+ */
+#ifndef ROCKSDB_LITE
+Status Configurable::SetOption(const DBOptions& db_opts,
+                               const std::string& name,
+                               const std::string& value,
+                               bool input_strings_escaped, bool* found_option) {
+  // By the time this method is called, the name is the short name (no prefix)
+  // and the value is unescaped.
+  Status status;
+  *found_option = false;
+  // Look for the name in all of the registered option maps until it is found
+  for (auto iter : options_) {
+    const auto& opt_map = iter.second.second;
+    void* opt_ptr = iter.second.first;
+    // Look up the value in the map
+    auto opt_iter = FindOption(name, opt_map);
+    if (opt_iter != opt_map.end()) {
+      const auto& opt_info = opt_iter->second;
+      *found_option = true;
+      if (name == opt_iter->first) {
+        status = ParseOption(opt_info, db_opts, opt_ptr, name, value,
+                             input_strings_escaped);
+      } else {
+        status = ParseOption(opt_info, db_opts, opt_ptr,
+                             name.substr(opt_iter->first.size() + 1), value,
```

Comment:
If `name` is "test.simple.int", and `opt_iter->first` is "int", then `name.substr(opt_iter->first.size() + 1)` does not seem right.

Reply:
I do not think that can happen.  If name is "test.simple.int", then opt_iter->first would either be "test.simple.int" (match completely) or "test" (matches to the first ".").  This case is handling the latter, and would pass "simple.int" to the next layer to be matched.

Also note that any options prefix is stripped off before this point.  So if the class had an option prefix of "test.simple" and the names passed in originally are "test.simple.int", test.simple.struct.double", and "test.bool", then the options processed here would be "int", "struct.double", and "test.bool" as the prefix would have already been removed.

Reply:
According to `simple_options_info` in configurable_test.cc, this can happen. Specifically in the example in configurable.cc, the opt_iter->first is "int", "bool", etc. and I believe the name can be "test.simple.int" or "test.simple.bool" if long form is used.

Reply:
Please disregard my previous comment about whether it can happen or not: we both agree it can happen.